### PR TITLE
Assignment2-ColtonC

### PR DESCRIPTION
**Pull request by Cronk77**

### CODE REVIEW

File: Logger/FileLogger.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    public string PathToFile {get; set; }
+
+    public FileLogger(string pathToFile)
+    {
+        PathToFile = pathToFile;
+    }
+
+    override public void Log(LogLevel logLevel, string message)
+    {
+        string currentDateAndTime = DateTime.Now.ToString("MM/d/yyy HH:mm:ss tt");
```
Comment (pja6) [inline Logger/FileLogger.cs:18]: I believe the way we resolved this error was with the following:
```suggestion
            string currentDateTime = DateTime.Now.ToString("MM/dd/yyyy hh:mm:ss tt", CultureInfo.CurrentCulture);
```

File: Logger/Logger.csproj [hunk 2]
```diff
@@ -2,7 +2,8 @@
 
   <PropertyGroup>
     <TargetFramework>netstandard2.0</TargetFramework>
-    <LangVersion>10.0</LangVersion>
+    <LangVersion>12.0</LangVersion>
+	<EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
```
Comment (pja6) [inline Logger/Logger.csproj:6]: Not sure if you enabled them another way but you could also add these here
```suggestion
	<EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
	<Nullable>enable</Nullable>
	<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
         <EnableNETAnalyzers>true</EnableNETAnalyzers>
         <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
         <TargetFramework>net8.0</TargetFramework>
```

**Review by pja6 — COMMENTED** [review]
There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ✔
Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
The current date/time ✔
The name of the class that created the logger ✔
The log level ✔
The message ✔

The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ✔
If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
Inside of BaseLoggerMixins implement extension methods on BaseLogger for
Error, ✔
Warning, ✔
Information, and ✔
Debug. ✔ 

Use the nameof() operator when identifying the class name to the logger ✔
Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ❌
Ensure that you turn on code analysis (EnableNETAnalyzers) ❌
Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ❌
Ensure that you turn on EnforceCodeStyleInBuild ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Turn on Nullability (Nullable) ❌

File: Logger/FileLogger.cs (hunk 1 above)
Comment (MattA41) [reply Logger/FileLogger.cs:18]: I think the formatting of the DateTime was wrong. I didn't have to use CultureInfo.CurrentCulture in my build

File: Logger.Tests/FileLoggerTests.cs [hunk 3]
```diff
@@ -5,5 +5,5 @@ namespace Logger.Tests;
 [TestClass]
 public class FileLoggerTests
 {
-
```
Comment (MattA41) [inline Logger.Tests/FileLoggerTests.cs:8]: Is there a reason that you did not do ant unit tests? this was to run tests on the methods in the FileLogger class

File: Logger.Tests/LogFactoryTests.cs [hunk 4]
```diff
@@ -1,10 +1,9 @@
 ﻿using Microsoft.VisualStudio.TestTools.UnitTesting;
-using System.IO;
 
 namespace Logger.Tests;
 
 [TestClass]
 public class LogFactoryTests
 {
-
```
Comment (MattA41) [inline Logger.Tests/LogFactoryTests.cs:9]: We are learning test-driven development, this file is for testing the methods in the LogFactory class. They are very helpful to making sure everything works especially when we are writing a library and cannot run the code to see if it works

File: Logger.Tests/Logger.Tests.csproj [hunk 5]
```diff
@@ -1,19 +1,24 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
 
-  <PropertyGroup>
-    <TargetFramework>net6.0</TargetFramework>
-    <IsPackable>false</IsPackable>
-  </PropertyGroup>
+	<PropertyGroup>
+		<TargetFramework>net8.0</TargetFramework>
```
Comment (MattA41) [inline Logger.Tests/Logger.Tests.csproj:4]: I added the other things we needed on like treating warnings as errors, enable net analyzers, code analysis treat warnings as error, and nullable. Having these on would have helped you catch errors and have your code build in the git build

File: Logger/BaseLoggerMixins.cs [hunk 6]
```diff
@@ -1,6 +1,47 @@
-﻿namespace Logger;
+﻿# nullable enable
+﻿using System;
+
+namespace Logger;
 
 public static class BaseLoggerMixins
```
Comment (MattA41) [inline Logger/BaseLoggerMixins.cs:6]:
```suggestion
public static class BaseLoggerMixins : BaseLogger
```
BaseLoggerMixins needs to extend BaseLogger

File: Logger/LogFactory.cs [hunk 7]
```diff
@@ -1,10 +1,27 @@
...
 
 public class LogFactory
 {
-    public BaseLogger CreateLogger(string className)
+    private string? PathToFile;
+    public void ConfigureFileLogger(string pathToFile)
     {
-        
-        return null;
+        PathToFile = pathToFile;
```
Comment (MattA41) [inline Logger/LogFactory.cs:9]: this assignment is unreached as it's after the return so pathtofile will always be null

**Review by MattA41 — COMMENTED** [review]
- There is an existing `BaseLogger` class. It needs an **auto property** to hold a class name. This property should be set in the `LogFactory` using an **object initializer**. ✔
- Create a `FileLogger` that derives from `BaseLogger`. It should take in a path to a file to write the log message to. When its `Log` method is called, it should **append** messages on their own line in the file. The output should include all of the following:
  - The current date/time ✔
  - The name of the class that created the logger ✔
  - The log level ✔
  - The message ✔
  - The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
- The `LogFactory` should be updated with a new method `ConfigureFileLogger`. This should take in a file path and store it in a **private member**. It should use this when instantiating a new `FileLogger` in its `CreateLogger` method. ✔
- If the file logger has not be configured in the `LogFactory`, its `CreateLogger` method should return `null`. ✔
- Inside of `BaseLoggerMixins` implement **extension methods** on `BaseLogger` for
  - `Error`, ✔
  - `Warning`, ✔
  - `Information`, and ✔
  - `Debug`. ✔
  Each of these methods should take in a `string` for the message, as well as a **parameter array** of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the `BaseLogger.Log` method, by automatically supplying the appropriate `LogLevel`. These methods should throw an exception if the `BaseLogger` parameter is null. There are a couple example unit tests to get you started.
- Use the nameof() operator when identifying the class name to the logger ❌
- Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ❌
- Ensure that you turn on code analysis (EnableNETAnalyzers) ❌
- Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ❌
- Ensure that you turn on EnforceCodeStyleInBuild ❌
- Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional) ✔
- Turn on Nullability (`Nullable`) ❌
- **All of the above should be unit tested.** ❌

File: Logger/FileLogger.cs (hunk 1 above)
Comment (Cronk77) [reply Logger/FileLogger.cs:18]: Thanks!

File: Logger/Logger.csproj (hunk 2 above)
Comment (Cronk77) [reply Logger/Logger.csproj:6]: Thanks!

File: Logger/BaseLoggerMixins.cs (hunk 6 above)
Comment (Cronk77) [reply Logger/BaseLoggerMixins.cs:6]: extension methods not derive

File: Logger.Tests/FileLoggerTests.cs (hunk 3 above)
Comment (Cronk77) [reply Logger.Tests/FileLoggerTests.cs:8]: writing now

File: Logger.Tests/LogFactoryTests.cs (hunk 4 above)
Comment (Cronk77) [reply Logger.Tests/LogFactoryTests.cs:9]: writing now

File: Logger/FileLogger.cs (hunk 1 above)
Comment (StrangeRanger) [reply Logger/FileLogger.cs]: @pja6 @MattA41 That's a good point out. `CultureInfo.CurrentCulture` isn't necessary in this instance, but it's something to keep in mind when it comes to localization. Refer to the [following link](https://learn.microsoft.com/en-us/dotnet/api/system.globalization.cultureinfo.currentculture?view=net-8.0) for more information.

**Comment by StrangeRanger** [comment]
@pja6

> Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ❌
> Ensure that you turn on code analysis (EnableNETAnalyzers) ❌
> Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ❌
> Turn on Nullability (Nullable) ❌

When it comes to the analyzers/configuration-based requirements, make sure to provide a little description, such as which projects don't contain them.

Example:

> Turn on Nullability (Nullable) ❌ => Not set in the test project

File: Logger/BaseLoggerMixins.cs [hunk 8]
```diff
@@ -1,6 +1,71 @@
-﻿namespace Logger;
+﻿using System;
+
+namespace Logger;
 
 public static class BaseLoggerMixins
 {
-    
+    public static void Error(string message, BaseLogger? baseLogger, params string?[] args)
```
Comment (BenjaminMichaelis) [inline Logger/BaseLoggerMixins.cs:7]: These should be extension methods

File: Logger/BaseLogger.cs [hunk 9]
```diff
@@ -3,5 +3,5 @@
 public abstract class BaseLogger
 {
     public abstract void Log(LogLevel logLevel, string message);
+    public string? ClassName { get; set; }
```
Comment (BenjaminMichaelis) [inline Logger/BaseLogger.cs:6]: Why is this nullable?

File: Logger/BaseLoggerMixins.cs [hunk 10]
```diff
@@ -1,6 +1,71 @@
...
+            throw new ArgumentNullException($"{baseLogger} is a null argument");
+        }
+        if (args is null)
+        {
+            baseLogger.Log(LogLevel.Error, message);
+        }
+
+        #pragma warning disable CS8604 // This will not be null since the statment above handles it
+        string fullMessage = $"{message} {string.Join(" ", args)}";
+        #pragma warning restore CS8604
```
Comment (BenjaminMichaelis) [inline Logger/BaseLoggerMixins.cs:20]: You should be using string.Format here instead of join

File: Logger/BaseLoggerMixins.cs [hunk 11]
```diff
@@ -1,6 +1,71 @@
...
+            baseLogger.Log(LogLevel.Error, message);
+        }
+
+        #pragma warning disable CS8604 // This will not be null since the statment above handles it
+        string fullMessage = $"{message} {string.Join(" ", args)}";
+        #pragma warning restore CS8604 
+        baseLogger.Log(LogLevel.Error, fullMessage);
+    }
+
+    public static void Warning(string message, BaseLogger? baseLogger, params string[] args)
```
Comment (BenjaminMichaelis) [inline Logger/BaseLoggerMixins.cs:24]: Why is logger nullable? I expect to call this message, a logger is needed.

File: Logger/LogFactory.cs [hunk 12]
```diff
@@ -2,9 +2,29 @@
 
 public class LogFactory
 {
-    public BaseLogger CreateLogger(string className)
+    private string? _pathToFile;
```
Comment (BenjaminMichaelis) [inline Logger/LogFactory.cs:5]: this should use a property instead of a field

File: Logger/LogFactory.cs [hunk 13]
```diff
@@ -2,9 +2,29 @@
...
+    public void ConfigureFileLogger(string PathToFile)
+    {
+        this._pathToFile = PathToFile;
+    }
+    
+    public BaseLogger? CreateLogger(string className)
     {
-        
-        return null;
+        if (className == nameof(FileLogger))
```
Comment (BenjaminMichaelis) [inline Logger/LogFactory.cs:18]: This limits the only class to be able to be passed into be FileLogger. We should allow this to work for any class

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 14]
```diff
@@ -7,33 +7,122 @@ namespace Logger.Tests;
...
+        // Arrange
+
+        // Act
+        BaseLoggerMixins.Warning("", null);
+
+        // Assert
+    }
+
+
+    // public void Warning_WithData_LogsMessage()
```
Comment (BenjaminMichaelis) [inline Logger.Tests/BaseLoggerMixinsTests.cs:54]: Looks like there a lot of tests we are missing

**Pull request closed without merging**